model: simplify semantics of compression + decompression of batches

rockwotj opened the pull request:
It sort of doesn't make sense TBH to support this, and now there is a
duality with the decompress methods, which throw if trying to decompress
something that is uncompressed

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

rockwotj committed:
batch/compression: don't allow compressing with model::compression::none

There are a few things here, one now we require that batches passed into
compress_batch are marked as uncompressed (callers updated), otherwise
if you tried to *change* compression type this method would mess up the
bit flags for compression in the header (thankfully we never do this,
but we should shore up the API).

Secondly we disallow calling `compress_batch` if you want the batch to
be uncompressed.

Lastly we throw instead of vassert, as this doesn't seem worthy of a
vassert.

WillemKauf commented on src/v/model/batch_compression.cc:
```diff
@@ -72,15 +72,10 @@ model::record_batch maybe_decompress_batch_sync(const model::record_batch& b) {
-    }
+    vassert(
+      b.header().attrs.compression() == model::compression::none,
+      "Asked to compress a batch with `none` compression: {}",
```
Does this `vassert()` line make sense? We're checking the header of the batch and inferring that the compression type `c` is `none` if the header compression type is `none`?

rockwotj replied:
Could turn this into a throw. But we're basically asserting that you're not giving a pre-compressed batch. If `c == none` then we will throw in `to_compression_type`. Let me write some tests to make the behavior clear.

WillemKauf replied:
Right, that makes sense and it is a valuable check, it just didn't match my expectations from the PR title. It also feels like the `vassert()` line would be more truthful to the comparision being performed if it were

`"Asked to compress a batch that already has a compression flag set: {}`"

or something along those lines.

rockwotj replied:
Fixed in latest revision!

rockwotj replied:
I also clarified the commit message here. I will update the PR title

WillemKauf commented on src/v/model/batch_compression.cc:
```diff
@@ -94,15 +89,10 @@ compress_batch(model::compression c, model::record_batch b) {
-    }
+    vassert(
+      b.header().attrs.compression() == model::compression::none,
+      "Asked to compress a batch with `none` compression: {}",
```
Ditto

rockwotj replied:
Fixed in force push

WillemKauf reviewed (commented):
The PR title is

> don't allow compressing with model::compression::none

but the change is asserting that the incoming batch is uncompressed, not that the input compression type is not `none`?

rockwotj committed:
model/tests: port record_batch_test to gtest

rockwotj committed:
model/batch_compression: add tests for batch compression/decompression

rockwotj committed:
batch/compression: simplify batch decompression API

Previously we had two versions of this API, some of which too an r-value
and some that took an l-value, and the semantics of the function wrt
uncompressed batches changed based on that, which feels icky. Tighten up
the semantics here to be consistent.

WillemKauf commented on src/v/model/batch_compression.cc:
```diff
@@ -28,6 +28,9 @@ ::compression::type to_compression_type(model::compression c) {
     case model::compression::zstd:
         return ::compression::type::zstd;
     case model::compression::none:
+        throw std::runtime_error(
```
It seems weird that we want to throw from this function (which is only used to convert types from different subsystems) instead of checking and throwing in one of the compression functions

rockwotj replied:
Done in the compress method

WillemKauf commented on src/v/model/batch_compression.cc:
```diff
@@ -52,7 +55,7 @@ model::record_batch decompress_batch_sync(model::record_batch&& b) {
 
 model::record_batch maybe_decompress_batch_sync(const model::record_batch& b) {
-    if (unlikely(!b.compressed())) {
+    if (!b.compressed()) [[unlikely]] {
```
squash these changes into the first commit?

rockwotj replied:
done

WillemKauf reviewed (dismissed):
Nice!

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--30ce5e17-feb2-465d-a9f8-9ef8e547e0c8-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70898">build#70898</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:-------------------------|:---------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/70898#0198b13b-c8e1-44f4-8fc3-68d647168263 | FLAKY         | 20/21    | upstream reliability is '99.51219512195122'. current run reliability is '95.23809523809523'. drift is 4.2741 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->
